docs/saved_objects/saved_object_access_control_design_proposal.md

```diff
@@ -0,0 +1,216 @@
+
+For the [find](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/routes/find.ts) API which lists the saved objects based on different search conditions, we will need to construct a search filter that can be applied to the query when performing searches against database.
+
+OpenSearch Dashboards provided a couple approaches to intercept saved objects operations, such as using [ClientWrapper](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/scoped_client_provider.ts#L74) and [SavedObjectsClientFactory](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/scoped_client_provider.ts#L82) . We can leverage these interception mechanisms to simplify the privilege evaluation implementation.
+
+
+#### Where to store ACLs
+
+The ACL is a logically a sub-resource/attribute that is attached to OpenSearch Dashboards resources, and the same access control mechanism should apply to the ACL as the saved object it attaches to.  We can either embed the ACL as attribute to the saved object, or store it separately as the standalone record, depending on what database is used by OpenSearch Dashboards. e.g. in the case of using NoSQL database as OpenSearch Dashboards metadata store, such as using OpenSearch index, embed the ACL into the saved object document as an attribute can simplify the privilege evaluation. while if using relational DB as OpenSearch Dashboards metadata store, storing ACLs as separate data records will be simpler.
+
+### Sharing with saved object hierarchy
+
+There are generally 3 ways to share saved objects in this design
+
+* Grant principals permissions to access workspace
+    * Only workspace admin can perform this operation
+* Add saved objects to another workspace
+    * the user needs to have write permission in both source and destination workspaces
+* Grant principals permissions to access specific saved objects
+    * the user needs to have write permission to the saved objects they want to share
+
+If a user only have read permission on the objects, they will be able to copy the object to an workspace they have access to. But this is not a typical “sharing”, as they make new object instances instead of really sharing the original objects.
+
+Given OpenSearch Dashboards saved objects has dependency hierarchy, such as dashboard depends on visualizations, and visualization in tern depends on index pattern. In any of the sharing cases, if a saved objects is shared, all its direct and transitive dependencies will also be shared. For example, if a dashboard is shared, then all the visualizations, index patterns that are used in the dashboard will be shared.
```

Comment:
It's clear that saved object inherit the permissions from the workspace, but if I attach ACL to a dashboard object, should visualizations and index patterns **inherit** dashboard object's ACL? Or the same ACL should attached to all visualization and index pattern objects specifically?